task/tekton-catalog-publish/0.1/tekton-catalog-publish.yaml

```diff
@@ -0,0 +1,69 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: tekton-catalog-publish
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.17.0"
+    tekton.dev/tags: catalog, bundles
+    tekton.dev/displayName: "Publish a Tekton Catalog"
+spec:
+  description: >-
+    This task publishes a Tekton catalog as bundles to a container registry
+  workspaces:
+    - name: catalog
+      description: >-
+        A workspace that holds the catalog to be published. The catalog must be stored
+        in the root of the workspace mount point, and is must follow the
+        [Tetkon Catalog](https://github.com/tektoncd/catalog#catalog-structure) structure.
+    - name: dockerconfig
+      description: >-
+        An optional workspace that allows providing a .docker/config.json
+        file for tkn to access the container registry. The file should be placed at
+        the root of the Workspace with name config.json.
+      optional: true
+
+  params:
+    - name: TKN_IMAGE
+      type: string
+      description: tkn CLI container image to run this task
+      default: gcr.io/tekton-releases/dogfooding/tkn@sha256:015bb6018d5eaf4523a775f4955c85f68077099753661941d06843d5a61a86df
+    - name: REGISTRY
+      type: string
+      description: The registry where bundles are published to
+    - name: PATH
+      type: string
+      description: The path in the registry
+      default: tekton/catalog/tasks
+    - name: TAG
+      type: string
+      description: An optional extra tag. If provided, tasks are tagged with it too.
+      default: ""
+
+  steps:
+    - name: publish
+      image: "$(params.TKN_IMAGE)"
+      workingDir: "$(workspaces.catalog.path)"
+      env:
+        - name: REGISTRY
+          value: $(params.REGISTRY)
+        - name: REGISTRY_PATH
+          value: $(params.PATH)
+        - name: TAG
+          value: $(params.TAG)
+      script: |
+        set -e
+        TARGET="${REGISTRY}"
+        [[ "${REGISTRY_PATH}" != "" ]] && TARGET="${TARGET}/${REGISTRY_PATH}"
+        cd task
+        find . -type f -mindepth 3 -maxdepth 3 -name '*yaml' | while read aa; do
```

Comment:
This script has a lot of issues, and the best way to find them all is to run it through shellcheck.  The output from my version of shellcheck:

```

In script.sh line 1:
set -e
^-- SC2148: Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In script.sh line 5:
find . -type f -mindepth 3 -maxdepth 3 -name '*yaml' | while read aa; do
                                                             ^--^ SC2162: read without -r will mangle backslashes.


In script.sh line 6:
  FOLDER=$(dirname $aa)
                   ^-^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
  FOLDER=$(dirname "$aa")


In script.sh line 7:
  VERSION=$(basename $FOLDER)
                     ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
  VERSION=$(basename "$FOLDER")


In script.sh line 8:
  TASK=$(basename $(dirname $FOLDER))
                  ^----------------^ SC2046: Quote this to prevent word splitting.
                            ^-----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
  TASK=$(basename $(dirname "$FOLDER"))


In script.sh line 9:
  cat $aa | tkn bundle push ${TARGET}/${TASK}:${VERSION} -f- &
      ^-^ SC2086: Double quote to prevent globbing and word splitting.
      ^-^ SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.
                            ^-------^ SC2086: Double quote to prevent globbing and word splitting.
                                      ^-----^ SC2086: Double quote to prevent globbing and word splitting.
                                              ^--------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
  cat "$aa" | tkn bundle push "${TARGET}"/"${TASK}":"${VERSION}" -f- &


In script.sh line 11:
    $cat $aa | tkn bundle push ${TARGET}/${TASK}:${TAG} -f- &
    ^--^ SC2154: cat is referenced but not assigned (for output from commands, use "$(cat ...)" ).
         ^-^ SC2086: Double quote to prevent globbing and word splitting.
                               ^-------^ SC2086: Double quote to prevent globbing and word splitting.
                                         ^-----^ SC2086: Double quote to prevent globbing and word splitting.
                                                 ^----^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
    $cat "$aa" | tkn bundle push "${TARGET}"/"${TASK}":"${TAG}" -f- &

For more information:
  https://www.shellcheck.net/wiki/SC2148 -- Tips depend on target shell and y...
  https://www.shellcheck.net/wiki/SC2046 -- Quote this to prevent word splitt...
  https://www.shellcheck.net/wiki/SC2154 -- cat is referenced but not assigne...
```

After fixing all the shellcheck issues, I have a few questions remaining, I'll address those separately

```
set -e
TARGET="${REGISTRY}"
[ "${REGISTRY_PATH}" != "" ] && TARGET="${TARGET}/${REGISTRY_PATH}"
cd task
find . -type f -mindepth 3 -maxdepth 3 -name '*yaml' | while read -r aa; do
  FOLDER=$(dirname "$aa")
  VERSION=$(basename "$FOLDER")
  TASK=$(basename "$(dirname "$FOLDER")")
  tkn < "$aa" bundle push "${TARGET}/${TASK}:${VERSION}" -f-
  [ "${TAG}" != "" ] && \
    tkn < "$aa" bundle push "${TARGET}/${TASK}:${TAG}" -f-
done
```

Essentially
* all strings quoted properly
* `read` with `-r`
* `cat` and the typo `$cat` replaced with `<` redirection
* removed & and sleep 0.1 with simply waiting for the tkn bundle push command to complete (separate comment on that).

Reply:
Fixed all the issues.